network_simulator/statistics.py

```diff
@@ -4,6 +4,186 @@
+
+    # TODO(sharon): functions and subclasses/events (if any) goes here.
+
+
 class Statistics(object):
-    # TODO(sharon): implement.
-    pass
\ No newline at end of file
+    """
+    Intended to be owned by the main loop to record all statistics within
+    the network.
+
+    :ivar dict link_stats: a map from (end_1_id, end_2_id) to
+    :class:`.LinkStats`.
+    :ivar dict flow_stats: a map from flow ID to :class:`.FlowStats`.
+    :ivar dict host_stats: a map from host ID to :class:`.HostStats`.
+    """
+    def __init__(self, link_stats={}, flow_stats={}, host_stats={}):
+        self.link_stats = link_stats
+        self.flow_stats = flow_stats
+        self.host_stats = host_stats
+
+    def get_link_stats(self, link):
+        """
+        Check whether a link has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Link link: link to be determined.
+        """
+        # Create a tuple of the pair and check if it exists in the dict.
+        tup = (link.end_1_addr, link.end_2_addr)
+
+        if tup not in self.link_stats:
+            self.link_stats[tup] = LinkStats
+        return self.link_stats[tup]
+
+    def get_flow_stats(self, flow):
+        """
+        Check whether a flow has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Flow flow: flow to be determined.
+        """
+        if flow.id not in self.flow_stats:
+            self.flow_stats[flow.id] = FlowStats
+        return self.flow_stats[flow.id]
+
+    def get_host_stats(self, host):
+        """
+        Check whether a host has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Host host: host to be determined.
+        """
+        if host.address not in self.host_stats:
+            self.host_stats[host.address] = HostStats
+        return self.host_stats[host.address]
+
+    def link_packet_loss(self, link, curr_time):
+        """
+        Record a packet loss of a link.
+
+        :param link: link of action.
+        :param curr_time: simulation time of loss.
+        """
+        stats = self.get_link_stats(link)
+        # TODO(team): Is this the most efficient?
+        stats.buffer_occupancy.append((curr_time, link.buffer_occupancy))
+        stats.packet_loss_times.append(curr_time)
+
+    def link_packet_transmitted(self, link, curr_time):
+        """
+        Record a packet transmission of a link.
+
+        :param link: link of action.
+        :param curr_time: simulation time of transmission.
+        """
+        stats = self.get_link_stats(link)
+        stats.buffer_occupancy.append((curr_time, link.buffer_occupancy))
```

Comment:
Same as before, buffer occupancy changes should be recorded in a separate function, which will be called whenever the buffer size changes. The `link_packet_transmitted' function will only record transmission times and packet sizes of packets that were transmitted.